devdoc/thandle_requirements.md

```diff
@@ -0,0 +1,135 @@
+
+## Exposed API
+
+```c
+
+/*to be used as the type of handle that wraps T*/
+#define THANDLE(T)
+
+/*to be used in a header file*/
+#define THANDLE_TYPE_DECLARE(T)
+
+/*to be used in a .c file*/
+#define THANDLE_TYPE_DEFINE(T)
+
+```
+
+## THANDLE(T)
+
+```c
+#define THANDLE(T) 
+```
+
+`THANDLE(T)` introduces a new incomplete type that can be used with all subsequent APIs constructed with `THANDLE`. `T` is a type that was introduced by `MU_DEFINE_STRUCT` in a .c file and was wrapped using `THANDLE_TYPE_DEFINE(T)` in the .c file.
+
+The type is const - so the following incorrect assignment fails at compile time:
+
+```c
+    THANDLE(T) one = t_create(...);
+    THANDLE(T) two;
+    two = one; /*compiler error*/
+```
+
+## THANDLE_TYPE_DECLARE(T)
+```c
+#define THANDLE_TYPE_DECLARE(T)
+```
+
+`THANDLE_TYPE_DECLARE` introduces several functions that can be used with `THANDLE(T)` type. These are `THANDLE_DEC_REF(T)`, `THANDLE_INC_REF(T)`, `THANDLE_ASSIGN(T)`, `THANDLE_INITIALIZE(T)`.
+
+###  THANDLE_DEC_REF(T)
+```c
+MOCKABLE_FUNCTION(, void, THANDLE_DEC_REF(T), THANDLE(T), t);
+```
+
+`THANDLE_DEC_REF` decrements the reference count of `t` and frees memory if reference count reaches 0.
+
+**SRS_THANDLE_02_001: [** If `t` is `NULL` then `THANDLE_DEC_REF` shall return. **]**
+
+**SRS_THANDLE_02_002: [** `THANDLE_DEC_REF` shall decrement the ref count of `t`.  **]**
+
+**SRS_THANDLE_02_003: [** If the ref count of `t` reaches 0 then `THANDLE_DEC_REF` shall call `dispose` (if not `NULL`) and free the used memory.  **]**
+
+### THANDLE_INC_REF(T)
+```c
+MOCKABLE_FUNCTION(, void, THANDLE_INC_REF(T), THANDLE(T), t);
+```
+
+`THANDLE_INC_REF` increments the reference count of `t`.
+
+**SRS_THANDLE_02_004: [** If `t` is `NULL` then `THANDLE_INC_REF` shall return. **]**
+
+**SRS_THANDLE_02_005: [** `THANDLE_INC_REF` shall increment the reference count of `t`. **]**
+
+### THANDLE_ASSIGN(T)
+```c
+MOCKABLE_FUNCTION(, void, THANDLE_ASSIGN(T), THANDLE(T) *, t1, THANDLE(T), t2 );
+```
+
+`THANDLE_ASSIGN` does t1=t2. Both `t1` and `t2` are existing constructed handles.
+
+
+**SRS_THANDLE_02_006: [** If `t1` is `NULL` then `THANDLE_ASSIGN` shall return. **]**
+
+**SRS_THANDLE_02_007: [** If `*t1` is `NULL` and `t2` is `NULL` then `THANDLE_ASSIGN` shall return. **]**
+
+**SRS_THANDLE_02_008: [** If `*t1` is `NULL` and `t2` is not `NULL` then `THANDLE_ASSIGN` shall increment the reference count of `t2` and store `t2` in `*t1`. **]**
+
+**SRS_THANDLE_02_009: [** If `*t1` is not `NULL` and `t2` is `NULL` then `THANDLE_ASSIGN` shall decrement the reference count of `*t1` and store `NULL` in `*t1`. **]**
+
+**SRS_THANDLE_02_010: [** If `*t1` is not `NULL` and `t2` is not `NULL` then `THANDLE_ASSIGN` shall increment the reference count of `t2`, shall decrement the reference count of `*t1` and store `t2` in `*t1`. **]**
+
+###  THANDLE_INITIALIZE(T)
+```c
+ MOCKABLE_FUNCTION(, void, THANDLE_INITIALIZE(T), THANDLE(T) *, t1, THANDLE(T), t2 );
```

Comment:
>t1 [](start = 64, length = 2)

This might deserve a bit better naming since there is clearly a comment in ln. 97 warning that this is not the same as the lvalue in THANDLE_ASSIGN.

Might I suggest naming the THANDLE_ASSIGN args lvalue and rvalue.
And a suggestion for t1 here would be wrapped_type, while t2 could be out_handle.

You consider it...

Reply:
renamed these to be lvalue and rvalue.

---
In reply to: [385369765](https://github.com/Azure/azure-c-util/pull/16#discussion_r385369765) [](ancestors = 385369765)